### PR TITLE
IFTAQCAA-14 - test script to cover mobile apps links

### PR DESCRIPTION
**Pull request by dkrutenko**
Click through mobile apps links to see if they work

### CODE REVIEW

**Comment by Nikodych** [comment]
looks like I would review two PR's, I would wait until we merge first one, because this includes code from the first

**Comment by Nikodych** [comment]
let me know, when it would be synced with master

**Comment by dkrutenko** [comment]
merged

File: src/main/java/com/softserveinc/ita/utils/WindowTabUtil.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.softserveinc.ita.utils;
+
+import com.codeborne.selenide.SelenideElement;
+import lombok.experimental.UtilityClass;
+
+import static com.codeborne.selenide.Selenide.*;
+
+@UtilityClass
+public class WindowTabUtil {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/utils/WindowTabUtil.java:9]: BrowserTabHelper

File: src/main/java/com/softserveinc/ita/utils/WindowTabUtil.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    public static String getTitleFromSecondTab(SelenideElement selector) {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/utils/WindowTabUtil.java:10]: empty line after class declaration

Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/utils/WindowTabUtil.java:10]: all operations with selectors should be only in page objects or element util

such class should provide separate operations for the browser, take a look at Max's PR, calling methods for switching between tabs from class, and calling in test from page method for getting title

Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/utils/WindowTabUtil.java:10]: create common page object for playmarkets with methods getAndroidAppTitle and ios

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 3]
```diff
@@ -61,4 +61,31 @@ public void verifyLanguageSwitchingTest() {
...
                 .isTrue();
     }
+
+    @DataProvider
+    public Object[][] mobileAppLinks() {
+        return new Object[][]{
+                {"Google Play", "AppStore", "ROZETKA"}};
+    }
+
+    @Test(dataProvider = "mobileAppLinks")
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:71]: I thinks here not necessary to use data provider, just create variables in test

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/utils/WindowTabUtil.java (hunk 2 above)
Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/utils/WindowTabUtil.java:10]: Max's pr is merged

File: src/main/java/com/softserveinc/ita/pageobjects/BasePage.java [hunk 4]
```diff
@@ -91,28 +91,23 @@ public T closeAdvertisingBannerIfDisplayed() {
         return (T) this;
     }
 
+    @Step("BasePage: Clicked on header menu")
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/BasePage.java:94]: Clicked on header menu button

File: src/main/java/com/softserveinc/ita/pageobjects/HomePage.java [hunk 5]
```diff
@@ -50,9 +50,9 @@ public HomePage openSocialMediaPage(String mediaName) {
 
         return this;
     }
-  
-   public List<String> getTitlesFromListOfLastViewedProducts() {
 
+    @Step("HomePage: Get titles list of last viewed products")
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/HomePage.java:54]: for get methods steps not neeeded

File: src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java [hunk 6]
```diff
@@ -7,9 +7,9 @@
 import static com.codeborne.selenide.Selenide.$x;
 import static java.time.Duration.ofSeconds;
 
-//TODO: move methods not related to this page to other page objects
 public class ProductPage extends BasePage<ProductPage> {
 
+    @Step("ProductPage: Received the price of the first product")
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:12]: same

File: src/main/java/com/softserveinc/ita/pageobjects/SearchResultPage.java [hunk 7]
```diff
@@ -15,6 +15,8 @@
 
 public class SearchResultPage extends BasePage<SearchResultPage> {
 
+
+    @Step("SearchResultPage: Got list with filtered goods by 'productName'")
```
[the remark this answers is not in the document]
Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/pageobjects/SearchResultPage.java:19]: and for get and boolean methods steps not needed

File: src/main/java/com/softserveinc/ita/utils/BrowserTabHelper.java [hunk 8]
```diff
@@ -1,18 +1,18 @@
...
 import lombok.experimental.UtilityClass;
 
 import static com.codeborne.selenide.Selenide.*;
 
 @UtilityClass
-public class WindowTabUtil {
-    public static String getTitleFromSecondTab(SelenideElement selector) {
+public class BrowserTabHelper {
+
+    public static void switchToSecondWindow() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/utils/BrowserTabHelper.java:10]: add steps here

"Switched to second tab"

File: src/main/java/com/softserveinc/ita/utils/BrowserTabHelper.java [hunk 9]
```diff
@@ -1,18 +1,18 @@
...
         switchTo().window(1);
-        var title = selector.getText();
+    }
+
+    public static void closeSecondAndSwitchToFirstWindow() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/utils/BrowserTabHelper.java:14]: same

File: src/main/java/com/softserveinc/ita/pageobjects/HeaderMenuModal.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package com.softserveinc.ita.pageobjects;
+
+import io.qameta.allure.Step;
+
+import static com.codeborne.selenide.Selenide.$x;
+import static com.softserveinc.ita.utils.BrowserTabHelper.*;
+
+public class HeaderMenuModal {
+
+    @Step("HeaderMenuModal: Switched to second tab and got the title from PlayMarket. After - switched back to the first tab")
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/HeaderMenuModal.java:10]: step not needed for get/boolean methods

File: src/main/java/com/softserveinc/ita/pageobjects/HeaderMenuModal.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    public String getAndroidAppTitle() {
+        switchToSecondWindow();
+        var androidTitle = $x("//*[@class='AHFaub']//ancestor::span").getText();
+        closeSecondAndSwitchToFirstWindow();
+
+        return androidTitle;
+    }
+
+    @Step("HeaderMenuModal: Switched to second tab and got the title from AppStore. After - switched back to the first tab")
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/HeaderMenuModal.java:19]: same

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 12]
```diff
@@ -53,6 +53,28 @@ public void verifyCategoryTransitionThroughDropdownTest(String title) {
...
+        return new Object[][]{
+                {"Facebook"},
+                {"Telegram"}
+        };
+    }
+
+    @Test(dataProvider = "rozetkaSocialMedia")
+    public void verifySocialMediaLinks(String socialMediaName) {
+        homePage.openSocialMediaPage(socialMediaName);
+        var expectedUrl = SocialMedia.valueOf(socialMediaName.toUpperCase()).getSocialMediaLink();
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java]: implement such interface:
```
public interface IUiTextEnum {

    static <T extends Enum<T> & IUiTextEnum> T fromValue(String name, Class<T> enumType) {
        EnumSet<T> all = allOf(enumType);

        for (T aEnum : all) {
            if (aEnum.name().equalsIgnoreCase(replace(name, SPACE, "_"))) {
                return aEnum;
            }
        }
        throw new IllegalArgumentException("Unknown " + enumType.getSimpleName() + ": " + name);
    }
}
```

create method in enum SocialMedia after implementing interface above

```
public static SocialMedia getSocialMediaBy(@NonNull String name) {
        return fromValue(name, SocialMedia.class);
    }
```

Comment (Nikodych) [reply src/test/java/com/softserveinc/ita/HomePageTest.java]: and after this you will be able replace this "SocialMedia.valueOf(socialMediaName.toUpperCase())" by "getSocialMediaBy(socialMediaName)"

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/test/java/com/softserveinc/ita/HomePageTest.java (hunk 12 above)
Comment (dkrutenko) [reply src/test/java/com/softserveinc/ita/HomePageTest.java]: this method was in master already, social media links was made by Maks

Comment (Nikodych) [reply src/test/java/com/softserveinc/ita/HomePageTest.java]: ok, probably create task on this and add my comments to description

**Review by Nikodych — APPROVED** [review]

**Pull request merged**